README.md
Time Zone List

a standard time zone list

## field

```json
{
  "list": "timezone list",
  "incompatible": "Time zone not currently supported",
  "updateTime": "last updated time"
}
```

```json
{
  "code": "code	name of country",
  "coordinates":"latitude and longitude ",
  "iana": " Internet Assigned Numbers Authority",
  "location": "country, territory, area, or subdivision",
  "offset": "the offset in milliseconds between the time zone and UTC time",
  "shortGmt": "short GMT",
  "longGmt": "long GMT",
  "shortNonLocation": "short specific non-location format",
  "longNonLocation": "long specific non-location format"
}
```


CHANGELOG.md
# Changelog

## [0.3.0](https://github.com/dev-standard/timezone/compare/0.2.0...0.3.0) (2022-12-13)

## [0.2.0](https://github.com/dev-standard/timezone/compare/0.1.1...0.2.0) (2022-12-10)


### Features

* support i18n ([c161099](https://github.com/dev-standard/timezone/commit/c1610999721d9c205ae9d57d8a2995f1419d8757))
* support intruoduce with esm ([185d9b4](https://github.com/dev-standard/timezone/commit/185d9b40c449c0b01129768893da3348b96b37a2))


### Bug Fixes

* rm .babelrc ([08eaef9](https://github.com/dev-standard/timezone/commit/08eaef900e7bcc82bf9ec979d1e5b3866e9294a7))

### 0.1.1 (2022-12-10)


### Features

* add data ([de0ed9c](https://github.com/dev-standard/timezone/commit/de0ed9c6cc53a1d2ed173c7990d7c350ddc1291d))
* timezone list ([235a082](https://github.com/dev-standard/timezone/commit/235a0825f7ad2fd95d4ff9f6cb80c40648e7fa18))


### Bug Fixes

* remove pushArgs at .release-it.json ([b531da9](https://github.com/dev-standard/timezone/commit/b531da9096bdf6d022084e4ed831e8b369539731))
* version ([1e53493](https://github.com/dev-standard/timezone/commit/1e53493a3a403e96b7434fb9fed26f20065f1d67))

src/refresh.js
/*
 * @Date: 2022-12-10 00:12:09
 * @LastEditors: liting luz.liting@gmail.com
 * @LastEditTime: 2022-12-10 11:53:25
 * @FilePath: /timezone/src/refresh.js
 */
import fs from 'fs/promises'
import path from 'path'
import chalk from 'chalk'

const countryCodeEntry = path.resolve(process.cwd(), './src/data/iso3166.tab')
const entry = path.resolve(process.cwd(), './src/data/zone1970.tab')
const output = path.resolve(process.cwd(), 'timezone.json')

let countryCodeEntryData = new Map()

const clearOldData = async () => {
  try {
    await fs.rm(output, { force: true, maxRetries: 3, recursive: true, retryDelay: 100 })
    console.log(chalk.green('rm old timezone.json successful'))
  } catch (err) {
    console.error(err)
  }
}

const getCountryCodeEntryData = async () => {
  try {
    const data = await fs.readFile(countryCodeEntry, 'utf-8')
    const lines = data.split('\n').filter(line => !line.startsWith('#'))
    countryCodeEntryData.clear()
    lines.forEach(line => {
      const [key, contry] = line.split(/\s+/u)
      countryCodeEntryData.set(key, contry)
    })
    console.log(chalk.green('get contry code successful'))
  } catch (err) {
    console.error(err)
  }
}

const refresh = async () => {
  try {
    const data = await fs.readFile(entry, 'utf-8')
    const lines = data.split('\n').filter(line => !line.startsWith('#'))
    const timezoneData = {
      data: [],
      updateTime: Date.now(),
    }
    lines.forEach(line => {
      const [codes, coordinates, tz, comments] = line.split('\t')
      codes.split(',').forEach(code => {
        timezoneData.data.push({
          code,
          comments: comments ?? '',
          coordinates,
          tz,
          contry: countryCodeEntryData.get(code) ?? '',
        })
      })
    })
    fs.writeFile(output, JSON.stringify(timezoneData), 'utf-8')
  } catch (err) {
    console.error(err)
  }
}

const run = async () => {
  await Promise.all([clearOldData(), getCountryCodeEntryData()])
  refresh()
}

run()


src/index.js
/*
 * @Date: 2022-12-10 00:12:09
 * @LastEditors: liting luz.liting@gmail.com
 * @LastEditTime: 2022-12-14 00:13:04
 * @FilePath: /timezone/src/index.js
 */
import chalk from 'chalk'
import fs from 'fs/promises'
import glob from 'glob'
import path from 'path'
import { formatInTimeZone, getTimezoneOffset } from 'date-fns-tz/esm'
import { datefnsLocales, formatLocaleKey } from './utils/index.js'
import {sortBy} from 'lodash-es'

const localesEntry = path.join(process.cwd(), '/node_modules/date-fns/esm/locale/')
const countryCodeEntry = path.resolve(process.cwd(), './src/data/iso3166.tab')
const entry = path.resolve(process.cwd(), './src/data/zone1970.tab')
const outputDir = path.join(process.cwd(), './list')

const locales = []
const locationCodeEntryData = new Map()

const clearOldData = async () => {
  try {
    await fs.rm(outputDir, { force: true, maxRetries: 3, recursive: true, retryDelay: 100 })
    console.log(chalk.green('clear old data successful'))
  } catch (err) {
    console.error(err)
  }
}

const getLocales = async () => {
  glob(
    '!(_)*/',
    {
      cwd: localesEntry,
      root: localesEntry,
      nomount: false,
    },
    (err, files) => {
      if (err) {
        console.error(err)

        return
      }

      files.forEach(async file => {
        locales.push({
          locale: formatLocaleKey(file),
          outputFileName: `${file.slice(0, -1)}.json`,
        })
      })
    }
  )
}

const getCountryCodeEntryData = async () => {
  try {
    const data = await fs.readFile(countryCodeEntry, 'utf-8')
    const lines = data.split('\n').filter(line => !line.startsWith('#'))
    locationCodeEntryData.clear()

    lines.forEach(line => {
      const [key, location] = line.split(/\s+/u)
      locationCodeEntryData.set(key, location)
    })
    console.log(chalk.green('get location code successful'))
  } catch (err) {
    console.error(err)
  }
}

const refresh = async () => {
  try {
    const data = await fs.readFile(entry, 'utf-8')
    const lines = data.split('\n').filter(line => !line.startsWith('#'))
    const timezoneLocalesMap = new Map()

    lines.forEach(line => {
      const [codes, coordinates, iana] = line.split('\t')

      codes.split(',').forEach(code => {
        locales.forEach(({ locale }) => {
          if (!timezoneLocalesMap.get(locale)) {
            timezoneLocalesMap.set(locale, {
              list: [],
              incompatible: [],
              updateTime: Date.now(),
            })
          }

          try {
            const format = formatInTimeZone(new Date(), iana, 'O...OOO|OOOO|z..zzz|zzzz', {
              locale: datefnsLocales[locale],
            }).split('|')
            // const format = formatInTimeZone(new Date(), iana, 'O...OOO|OOOO|z..zzz|zzzz').split('|')

            timezoneLocalesMap.get(locale).list.push({
              code,
              coordinates,
              iana,
              location: locationCodeEntryData.get(code) ?? '',
              offset: getTimezoneOffset(iana),
              shortGmt: format[0].split('...')[0],
              longGmt: format[1],
              shortNonLocation: format[2].split('..')[0],
              longNonLocation: format[3],
            })
          } catch (err) {
            // console.error(err)
            timezoneLocalesMap.get(locale).incompatible.push({
              code,
              coordinates,
              iana,
              location: locationCodeEntryData.get(code) ?? '',
              err: JSON.parse(JSON.stringify(err, ['name', 'message'], 2)),
            })
          }
        })
      })
    })
    await fs.mkdir(outputDir)

    locales.forEach(locale => {
      timezoneLocalesMap.get(locale.locale).list = sortBy(timezoneLocalesMap.get(locale.locale).list, ['offset', 'iana', 'contry'])
      timezoneLocalesMap.get(locale.locale).incompatible = sortBy(timezoneLocalesMap.get(locale.locale).incompatible, ['iana', 'contry'])
      fs.writeFile(
        path.resolve(outputDir, locale.outputFileName),
        JSON.stringify(timezoneLocalesMap.get(locale.locale), null, 2),
        'utf-8'
      )
    })
  } catch (err) {
    console.error(err)
  }
}

const run = async () => {
  await Promise.all([clearOldData(), getCountryCodeEntryData(), getLocales()])
  refresh()
}

run()
